src/interpreter/blockchain.rs

```diff
@@ -308,4 +309,64 @@ where
 
         self.inc_pc()
     }
+
+    pub(crate) fn message_output(&mut self, a: Word, b: Word, c: Word, d: Word) -> Result<(), RuntimeError> {
+        if b > self.params.max_message_data_length
+            || b > MEM_MAX_ACCESS_SIZE
+            || a > VM_MAX_RAM - b - Bytes32::LEN as Word
+        {
+            return Err(PanicReason::MemoryOverflow.into());
+        }
+
+        let idx = c;
+        let amount = d;
+
+        // Safety: checked len
+        let recipient = unsafe { Address::from_slice_unchecked(&self.memory[a as usize..a as usize + Address::LEN]) };
+        if recipient == Address::zeroed() {
+            return Err(PanicReason::ZeroedMessageOutputRecipient.into());
+        }
+
+        let offset = self
+            .tx
+            .output_offset(c as usize)
+            .map(|ofs| ofs + self.tx_offset())
+            .ok_or(PanicReason::OutputNotFound)?;
+
+        // halt with I/O error because tx should be serialized correctly into vm memory
+        let output = Output::from_bytes(&self.memory[offset..])?;
+
+        // amount isn't checked because we are allowed to send zero balances with a message
+        if !matches!(output, Output::Message { recipient, .. } if recipient == Address::zeroed()) {
+            return Err(PanicReason::NonZeroMessageOutputRecipient.into());
+        }
+
+        // validations passed, perform the mutations
+
+        // subtract the balance only if there is an amount set
+        if amount > 0 {
+            if self.is_external_context() {
+                self.base_asset_balance_sub(amount)?;
+            } else {
+                self.external_asset_balance_sub(amount)?;
+            }
+        }
+
+        let fp = self.registers[REG_FP] as usize;
+        let txid = self.tx_id();
+        let data = a as usize + Address::LEN;
```

Comment:
Maybe this should use `Bytes32::LEN` like on line 316 for consistency.

Reply:
True, they should be the same. The concrete type is in fact `Address`, instead of raw bytes. Better to use that so its clear what is expected of this register:

https://github.com/FuelLabs/fuel-vm/pull/159/commits/7bd578949cb2d81e1bd84388532d33e523473636